src/core/server/http/http_service.ts

```diff
@@ -150,4 +169,28 @@ export class HttpService implements CoreService<HttpServiceSetup, HttpServiceSta
     await Promise.all([...this.secondaryServers.values()].map(s => s.stop()));
     this.secondaryServers.clear();
   }
+
+  private async runNotReadyServer(config: HttpConfig) {
+    const httpServer = new HttpServer(this.log);
+    const { server } = await httpServer.setup(config);
+    this.notReadyServer = server;
+    // use hapi server while Kibana ResponseFactory doesn't allow specifying custom headers
+    // https://github.com/elastic/kibana/issues/33779
+    this.notReadyServer.route({
+      path: '/{p*}',
+      method: '*',
+      handler: (req, responseToolkit) => {
+        this.log.debug(`Kibana server is not ready yet ${req.method}:${req.url}.`);
+
+        // If server is not ready yet, because plugins or core can perform
+        // long running tasks (build assets, saved objects migrations etc.)
+        // we should let client know that and ask to retry after 30 seconds.
+        return responseToolkit
+          .response('Kibana server is not ready yet')
+          .code(503)
```

Comment:
Status codes in the 500 range denote that a server error has occurred, but that isn't the case here. For "not ready yet" messages, in the past I've typically used a 202 with `Retry-After`, but I don't have strong opinions here. I just don't want to conflate a server error with not ready yet.

Reply:
we can reconsider this in a separate issue

Reply:
Actually `503 + Retry-After` is 

> 10.5.4 503 Service Unavailable
> The server is currently unable to handle the request due to a temporary overloading or maintenance of the server. The implication is that this is a temporary condition which will be alleviated after some delay. If known, the length of the delay MAY be indicated in a Retry-After header. If no Retry-After is given, the client SHOULD handle the response as it would for a 500 response.

Looks like it fits our use case, but we can discuss that in a separate issue, yes.